repo/blob/gcs/gcs_storage.go

```diff
@@ -264,12 +298,18 @@ func New(ctx context.Context, opt *Options, isCreate bool) (blob.Storage, error)
 		return nil, errors.New("bucket name must be specified")
 	}
 
-	gcs := &gcsStorage{
+	st := &gcsStorage{
 		Options:       *opt,
 		storageClient: cli,
 		bucket:        cli.Bucket(opt.BucketName),
 	}
 
+	fmt.Println("Provo il PIT")
```

Comment:
in general the Println's can go

Reply:
I'll look into it Today, see if I notice anything

Reply:
main issue i had is that GCS doesn't allow sorting object versions by creation time so i had to list all versions, sort manually by creation time, cycle the result and keep the newestthan (or older then). but im not a Go programmer, i had huge issues trying to sort and debug this thing. i have an apparently working version on my laptop but i'm not confident on why It works 🤣

(probably on Discord i can explain better)

Reply:
So a couple notes from looking at it.
- blob versions are automatically sorted from oldest to most recently created so no need to add sorting anywhere
- `IsDeleteMarker` needs to consider not only if `Deleted` is set but also if it was deleted after the PIT e.g. `!oi.Deleted.IsZero() && (gcs.PointInTime == nil || oi.Deleted.Before(*gcs.PointInTime))`. For convenience you can change `infoToVersionMetadata` to be a method instead of func
- `newestAtUnlessDeleted` needs to return `!v.IsDeleteMarker` as mentioned 
- `getOlderThan` needs `return vs[:i]` like Azure since they're ascending 

If you make these changes then PIT will work as expected

Not really to PIT in particular:
I guess better if Timestamp is .Created rather than .Updated

Reply:
> blob versions are automatically sorted from oldest to most recently created so no need to add sorting anywhere

Not with Google Cloud Storage, that's why i wasn't unable to get this working. Google doesn't have a stable sort currently.

Reply:
Something link this seem to fix the sorting and also fix getting ALL versions before PIT and the LATEST version BEFOR PIT (i did some test on simple arrays, not on object from google cloud). An unsorted array like `[]int{2, 7, 14, 4, 9, 22, 23, 6}` returns `[2,4,6]` from `getOlderThan` and `6` from `newestAtUnlessDeleted` supposing a PIT set to `7`
```
// newestAtUnlessDeleted returns the last version in the list older than the PIT.
func newestAtUnlessDeleted(vs []versionMetadata, t time.Time) (v versionMetadata, found bool) {
	fmt.Printf("newestAtUnlessDeleted %s\n", t)

	vs = getOlderThan(vs, t)

	if len(vs) == 0 {
		return versionMetadata{}, false
	}

	v = vs[len(vs)-1]
	fmt.Printf("newestAtUnlessDeleted del pit %s deleted=%t\n", v.Timestamp, v.IsDeleteMarker)

	//return v, !v.IsDeleteMarker
	return v, true
}

// Removes versions that are newer than t. The filtering is done in place
// and uses the same slice storage as vs. Assumes entries in vs are in descending
// timestamp order.
func getOlderThan(vs []versionMetadata, t time.Time) []versionMetadata {
	fmt.Printf("getOlderThan %s \n", t)

	sort.Slice(vs, func(i, j int) bool { return vs[i].Timestamp.Before(vs[j].Timestamp) })

	for i := range vs {
		fmt.Printf("getOlderThan %s: vs[i].Timestamp: %s \n", t, vs[i].Timestamp)
		if vs[i].Timestamp.After(t) {
			fmt.Printf("  Skippo versione %s del %s. Troppo nuova.\n", vs[i].Version, vs[i].Timestamp)
			return vs[:i]
		}
	}

	return vs
}
```

Reply:
> > blob versions are automatically sorted from oldest to most recently created so no need to add sorting anywhere
> 
> Not with Google Cloud Storage, that's why i wasn't unable to get this working. Google doesn't have a stable sort currently.

Every time I call list blobs like this: `$HOME/go/bin/kopia blob list --prefix=untitled`
I get all 4 versions of the blob in ascending order at the point of "newestAtUnlessDeleted: PRE SORT"
If it wasn't stable then at least once it would vary unless you are sorting before `PRE SORT` also

Reply:
You can see the changes I mentioned [here](https://github.com/KastenMike/kopia/pull/10/files)

`return v, true` is not correct since it would return deleted blobs e.g. PIT=15:00, blob deleted 14:00

Reply:
I @'d you at the lines since I realised I didn't do them in a separate commit

Reply:
```
****found blob: untitled file -- om.BlobID: untitled file -- query.Prefix: untitled -- prefix: untitled
newestAtUnlessDeleted 2024-08-29 15:04:05 +0000 UTC with 6 versions
newestAtUnlessDeleted: PRE SORT: found version 1724921148084300 del 2024-08-29 08:45:48.088 +0000 UTC. deleted=true
newestAtUnlessDeleted: PRE SORT: found version 1724921200485667 del 2024-08-29 08:46:40.497 +0000 UTC. deleted=true
newestAtUnlessDeleted: PRE SORT: found version 1724921590145980 del 2024-08-29 08:53:10.156 +0000 UTC. deleted=true
newestAtUnlessDeleted: PRE SORT: found version 1724926790177147 del 2024-08-29 10:19:50.18 +0000 UTC. deleted=true
newestAtUnlessDeleted: PRE SORT: found version 1724935441257780 del 2024-08-29 12:44:01.268 +0000 UTC. deleted=true
newestAtUnlessDeleted: PRE SORT: found version 1724935889814920 del 2024-08-29 12:51:29.826 +0000 UTC. deleted=false
```

Reply:
> Every time I call list blobs like this: `$HOME/go/bin/kopia blob list --prefix=untitled`
> I get all 4 versions of the blob in ascending order at the point of "newestAtUnlessDeleted: PRE SORT"
> If it wasn't stable then at least once it would vary unless you are sorting before `PRE SORT` also

I'm 100% sure about this, i've opened, times ago a request on Google and they said that the sort is not stable. But it's easy to fix, just sort the versioning on Kopia side. There is a feature request opened for google to implement a custom sorting feature in the API.

Reply:
They might not have custom sorting, but we don't need it. Or manual sorting. You can delete your sort calls and create 10 versions of a blob and every list command will have them in ascending order at the `PRE SORT` log you added

Reply:
> You can see the changes I mentioned [here](https://github.com/KastenMike/kopia/pull/10/files)
> 
> `return v, true` is not correct since it would return deleted blobs e.g. PIT=15:00, blob deleted 14:00 even when it is the latest/only blob version

I've seen your branch. If it works, i can safely drop this PR and fork yours as mine was just a wip and not a working one.

Reply:
> They might not have custom sorting, but we don't need it. Or manual sorting. You can delete your sort calls and create 10 versions of a blob and every list command will have them in ascending order at the `PRE SORT` log you added

When i did this PR it wasn't stable, that's why i've added a sort function, trying to make that stable. Probably they have changed something .....